src/json-schema.js

```diff
@@ -26,58 +31,101 @@ function convertSubSchema(schema) {
   }
 
   if (schema.allOf) {
-    actualSchema.allOf = schema.allOf.map(convertSubSchema);
+    actualSchema.allOf = schema.allOf.map(s => convertSubSchema(s, references));
   }
 
   if (schema.anyOf) {
-    actualSchema.anyOf = schema.anyOf.map(convertSubSchema);
+    actualSchema.anyOf = schema.anyOf.map(s => convertSubSchema(s, references));
   }
 
   if (schema.oneOf) {
-    actualSchema.oneOf = schema.oneOf.map(convertSubSchema);
+    actualSchema.oneOf = schema.oneOf.map(s => convertSubSchema(s, references));
   }
 
   if (schema.not) {
-    actualSchema.not = convertSubSchema(schema.not);
+    actualSchema.not = convertSubSchema(schema.not, references);
   }
 
   // Array
 
   if (schema.items) {
     if (Array.isArray(schema.items)) {
-      actualSchema.items = schema.items.map(convertSubSchema);
+      actualSchema.items = schema.items.map(s => convertSubSchema(s, references));
     } else {
-      actualSchema.items = convertSubSchema(schema.items);
+      actualSchema.items = convertSubSchema(schema.items, references);
     }
   }
 
   if (schema.additionalItems && typeof schema.additionalItems === 'object') {
-    actualSchema.additionalItems = convertSubSchema(schema.additionalItems);
+    actualSchema.additionalItems = convertSubSchema(schema.additionalItems, references);
   }
 
   // Object
 
   if (schema.properties) {
     Object.keys(schema.properties).forEach((key) => {
-      actualSchema.properties[key] = convertSubSchema(schema.properties[key]);
+      actualSchema.properties[key] = convertSubSchema(schema.properties[key], references);
     });
   }
 
   if (schema.patternProperties) {
     Object.keys(schema.patternProperties).forEach((key) => {
-      actualSchema.patternProperties[key] = convertSubSchema(schema.patternProperties[key]);
+      actualSchema.patternProperties[key] =
+        convertSubSchema(schema.patternProperties[key], references);
     });
   }
 
   if (schema.additionalProperties && typeof schema.additionalProperties === 'object') {
-    actualSchema.additionalProperties = convertSubSchema(schema.additionalProperties);
+    actualSchema.additionalProperties = convertSubSchema(schema.additionalProperties, references);
   }
 
   return actualSchema;
 }
 
+function lookupReference(reference, root) {
+  const parts = reference.split('/');
+
+  if (parts[0] !== '#') {
+    throw new Error('Schema reference must start with document root (#)');
+  }
+
+  if (parts[1] !== 'definitions' || parts.length !== 3) {
```

Comment:
Does it need to be definitions? This can be anything, right?

Reply:
In theory, but for the Swagger document to pass the swagger schema you won't be able to place a schema in many other places.

It gets a lot more complex to support "anything" because we have to copy that reference into our schema. "definitions" is the place in Swagger 2 and also JSON Schema specification that is designed for referencing. The "definitions" keyword does not directly affect the validation result. For example, if I we're to reference some other part that could break the schema:

```
{
  "type": "array",
  "items": {
    "$ref": "#/maxItems"
  }
  "maxItems": { // placing maxItems here is invalid as maxItems has semantics in JSON Schema and shouldn't be an object.
    "type": "object"
  }
}
```

From JSON Schema spec:

> 9. Schema Re-Use With "definitions"
> 
> The "definitions" keywords provides a standardized location for schema authors to inline re-usable JSON Schemas into a more general schema. The keyword does not directly affect the validation result.
> 
> This keyword's value MUST be an object. Each member value of this object MUST be a valid JSON Schema. 

I would just support this case unless we see a others placing it elsewhere but I don't think anyone is doing that.